bwp91/homebridge-govee
# feat: Integrate LAN API Interface

Hey! 👋

I have created this PR that integrates the new LAN API Interface introduced by Govee in v5.2.0.

I am not a JavaScript developer, so there's definitely some refactoring remaining to be done but I think this progressed nicely. I will be marking this as DRAFT, as there is still some more work to be done but still creating the PR to get some feedback in the meantime.

### Use Case:
To improve communication speed, reliability and to to mitigate intermediary communications, we should be able to communicate with supported Govee devices on the local network. 

### Description
Govee has introduced a new LAN API Interface for some device models (firmware update required) with more to be updated. The documentation can be found here: https://app-h5.govee.com/user-manual/wlan-guide

Then when sending update commands, the code will check `lanModels` to see wether we should attempt sending commands over LAN for this device. If  commands can be sent over LAN and a supported command is attempted, the command will be sent and other methods will be ignored. Otherwise, the previous behaviour shouldn't be affected.

### Acceptance Criteria:
- Discover the devices on local network
- Be able to query a device's state on local network
- Be able to send supported commands through LAN
- Unsupported commands should fallback to any other previous method of communication (if available)

### Testing Steps:
1. Go to Govee app -> Select Supported Device -> Settings (gear Icon) -> Enable LAN Control
2. Check that the device appears in Homebridge logs and all supported commands work

### Tasks Remaining
 - [x] LAN devices discovery
 - [x] Standard commands (turn on/off, color change) implemented
 - [x] Receive device status and update internal state after a command was sent
 - [x] Prefer polling for device state on LAN compatible devices rather than API/HTTP
 - [x] Have logs respect debug settings
 - [x] Send update commands faster on LAN devices (i.e., controlInterval should be much lower in this case)
 - [x] Device discovery without relying on API? (I.e, be able to work fully offline without API/HTTP/AWS)
